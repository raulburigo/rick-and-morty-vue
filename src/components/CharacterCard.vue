<template>
  <v-card>
    <v-img
      :src="char.image"
      class="white--text align-end"
      gradient="to bottom, rgba(0,0,0,.1), rgba(0,0,0,.5)"
      cover
    >
    </v-img>
    <v-card-title v-text="char.name"></v-card-title>
    <v-card-subtitle>
      <span v-if="char.origin.name != 'unknown'">{{ char.origin.name }}</span>
      <span v-else>Origem desconhecida</span>
    </v-card-subtitle>

    <v-card-actions>
      <v-btn color="#3bb5c0" text> Epis√≥dios </v-btn>

      <v-spacer></v-spacer>

      <v-btn icon @click="show = !show">
        <v-icon>{{ show ? "mdi-chevron-up" : "mdi-chevron-down" }}</v-icon>
      </v-btn>
    </v-card-actions>

    <v-expand-transition>
      <div v-show="show" data-testid="episodes">
        <v-divider></v-divider>

        <v-card-text>
          <v-virtual-scroll
            :items="char.episode"
            :item-height="50"
            :bench="3"
            min-height="65"
            max-height="150"
          >
            <template v-slot="{ item }">
              <v-list-item two-line>
                <v-list-item-content>
                  <v-list-item-title>
                    {{ item.episode }}
                  </v-list-item-title>
                  <v-list-item-subtitle>
                    {{ item.name }}
                  </v-list-item-subtitle>
                </v-list-item-content>
              </v-list-item>
            </template>
          </v-virtual-scroll>
        </v-card-text>
      </div>
    </v-expand-transition>
  </v-card>
</template>

<script>
export default {
  name: "CharacterCard",

  props: { char: Object },

  data: function () {
    return {
      show: false,
    };
  },
};
</script>
