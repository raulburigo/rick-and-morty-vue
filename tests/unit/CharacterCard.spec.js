/* eslint-disable no-unused-vars */
import Vue from "vue";
import Vuetify from "vuetify";

import { createLocalVue, mount, shallowMount } from "@vue/test-utils";

import CharacterCard from "../../src/components/CharacterCard";

const localVue = createLocalVue();
Vue.use(Vuetify);

describe("CharacterCard.vue", () => {
  let vuetify;
  const rickCard = {
    id: 1,
    image: "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
    origin: {
      name: "Earth (C-137)",
    },
    name: "Rick Sanchez",
    episode: [
      {
        episode: "S01E01",
        name: "Pilot",
      },
      {
        episode: "S01E02",
        name: "Lawnmower Dog",
      },
    ],
  };
  const alienMortyCard = {
    id: 14,
    image: "https://rickandmortyapi.com/api/character/avatar/14.jpeg",
    origin: {
      name: "unknown",
    },
    name: "Alien Morty",
    episode: [
      {
        episode: "S01E10",
        name: "Close Rick-counters of the Rick Kind",
      },
    ],
  };

  beforeEach(() => {
    vuetify = new Vuetify();
  });

  it("should have a custom title and subtitle and match snapshot", () => {
    const wrapper = mount(CharacterCard, {
      localVue,
      vuetify,
      propsData: { char: rickCard },
    });

    expect(wrapper.html()).toMatchSnapshot();

    const title = wrapper.find(".v-card__title");

    expect(title.text()).toBe("Rick Sanchez");

    const subTitle = wrapper.find(".v-card__subtitle");

    expect(subTitle.text()).toBe("Earth (C-137)");
  });

  it("should have a subtitle 'Origem Desconhecida' when origin name is 'unknown'", () => {
    const wrapper = mount(CharacterCard, {
      localVue,
      vuetify,
      propsData: { char: alienMortyCard },
    });

    const subTitle = wrapper.find(".v-card__subtitle");

    expect(subTitle.text()).toBe("Origem desconhecida");
  });

  it("should have as many scroll items as episodes with custom title and subtitle", () => {
    const rickWrapper = mount(CharacterCard, {
      localVue,
      vuetify,
      propsData: { char: rickCard },
    });

    const scrollItems = rickWrapper.findAll(
      ".v-virtual-scroll__container > .v-virtual-scroll__item"
    );

    expect(scrollItems.length).toBe(2);

    const epTitles = rickWrapper.findAll(".v-list-item__title");
    expect(epTitles.at(0).text()).toBe("S01E01");
    expect(epTitles.at(1).text()).toBe("S01E02");

    const epSubTitles = rickWrapper.findAll(".v-list-item__subtitle");
    expect(epSubTitles.at(0).text()).toBe("Pilot");
    expect(epSubTitles.at(1).text()).toBe("Lawnmower Dog");

    const alienMortyWrapper = mount(CharacterCard, {
      localVue,
      vuetify,
      propsData: { char: alienMortyCard },
    });

    const newScrollItems = alienMortyWrapper.findAll(
      ".v-virtual-scroll__container > .v-virtual-scroll__item"
    );

    expect(newScrollItems.length).toBe(1);
  });

  it("should show/hide episodes when button clicked", async () => {
    const wrapper = mount(CharacterCard, {
      localVue,
      vuetify,
      propsData: { char: rickCard },
    });

    const button = wrapper.find(".v-card__actions > .v-btn--icon");
    const episodes = wrapper.find('[data-testid="episodes"]');

    expect(episodes.isVisible()).toBe(false);

    await button.trigger("click");
    expect(episodes.isVisible()).toBe(true);

    await button.trigger("click");
    expect(episodes.isVisible()).toBe(false);
  });
});
