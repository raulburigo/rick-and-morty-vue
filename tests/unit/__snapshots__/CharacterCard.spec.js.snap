// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCard.vue should have a custom title and subtitle and match snapshot 1`] = `
<div class="v-card v-sheet theme--light">
  <div class="v-image v-responsive white--text align-end theme--light" cover="">
    <transition-stub name="fade-transition" mode="in-out">
      <div class="v-image__image v-image__image--preload v-image__image--cover" style="background-position: center center;"></div>
    </transition-stub>
    <div class="v-responsive__content"></div>
  </div>
  <div class="v-card__title">Rick Sanchez</div>
  <div class="v-card__subtitle"><span>Earth (C-137)</span></div>
  <div class="v-card__actions"><button type="button" class="v-btn v-btn--text theme--light v-size--default" style="color: rgb(59, 181, 192); caret-color: #3bb5c0;"><span class="v-btn__content"> Epis√≥dios </span></button>
    <div class="spacer"></div> <button type="button" class="v-btn v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-down theme--light"></i></span></button>
  </div>
  <transition-stub name="expand-transition">
    <div data-testid="episodes" style="display: none;">
      <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
      <div class="v-card__text">
        <div class="v-virtual-scroll" style="min-height: 65px; max-height: 150px;">
          <div class="v-virtual-scroll__container" style="height: 100px;">
            <div class="v-virtual-scroll__item" style="top: 0px;">
              <div tabindex="-1" class="v-list-item v-list-item--two-line theme--light">
                <div class="v-list-item__content">
                  <div class="v-list-item__title">
                    S01E01
                  </div>
                  <div class="v-list-item__subtitle">
                    Pilot
                  </div>
                </div>
              </div>
            </div>
            <div class="v-virtual-scroll__item" style="top: 50px;">
              <div tabindex="-1" class="v-list-item v-list-item--two-line theme--light">
                <div class="v-list-item__content">
                  <div class="v-list-item__title">
                    S01E02
                  </div>
                  <div class="v-list-item__subtitle">
                    Lawnmower Dog
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </transition-stub>
</div>
`;
